packages/pouch-storage/src/PouchCommandHistoryTable.ts

```diff
@@ -13,16 +29,131 @@ export class PouchCommandHistoryTable
     super(`CommandHistoryStorage.${language}`);
   }
 
-  setSearch(search: string): void {
-    this.setFilters(
-      search
-        ? [
-            StorageUtils.makeFilterConfig([
-              StorageUtils.makeFilterItem('name', FilterType.contains, search),
-            ]),
-          ]
-        : []
-    );
+  /**
+   * Singleton cache for CommandHistoryStorageItem arrays keyed by databaseName.
+   * This allows us to share results across multiple instances of
+   * PouchCommandHistoryTable.
+   */
+  static cache: Record<
+    string,
+    Promise<CommandHistoryStorageItemFindRespose> | null
+  > = {};
+
+  private searchText = '';
+
+  setSearchDebounceTimeout?: number;
+
+  setSearch(searchText: string): void {
+    clearTimeout(this.setSearchDebounceTimeout);
+
+    // Debounced search to minimize querying the PouchDB db while user types
+    this.setSearchDebounceTimeout = window.setTimeout(() => {
+      log.debug('Clearing cache and setting search filters', searchText);
+      log.debug('db.name', this.db.name);
+
+      PouchCommandHistoryTable.cache[this.db.name] = null;
+
+      this.searchText = searchText;
+      this.setFilters(
+        searchText
+          ? [
+              StorageUtils.makeFilterConfig([
+                StorageUtils.makeFilterItem(
+                  'name',
+                  FilterType.contains,
+                  searchText
+                ),
+              ]),
+            ]
+          : []
+      );
+    }, 500);
+  }
+
+  dbUpdate(
+    event: PouchDB.Core.ChangesResponseChange<CommandHistoryStorageItem>
+  ): void {
+    log.debug('Clearing cache and refreshing data', event);
+
+    PouchCommandHistoryTable.cache[this.db.name] = null;
+
+    super.dbUpdate(event);
+  }
+
+  private async fetchData(
+    selector: PouchDB.Find.Selector
+  ): Promise<
+    PouchDB.Find.FindResponse<CommandHistoryStorageItem & PouchStorageItem>
+  > {
+    if (PouchCommandHistoryTable.cache[this.db.name]) {
+      log.debug('Fetching from cache', this.searchText, this.viewport);
+    } else {
+      log.debug('Fetching from PouchDB', this.searchText, this.viewport);
+
+      if (this.searchText) {
+        PouchCommandHistoryTable.cache[this.db.name] = this.db.find({
+          selector,
+          fields: ['id', 'name'],
+          limit: DISPLAY_COMMAND_HISTORY_LIMIT,
+          sort: [{ id: 'desc' }],
+        });
+      } else {
+        PouchCommandHistoryTable.cache[this.db.name] = this.db
+          .allDocs({
+            include_docs: true,
+            descending: true,
+            limit: DISPLAY_COMMAND_HISTORY_LIMIT + 1, // +1 for an extra row returned by allDocs
```

Comment:
So we're still adding to the storage every time with `.put()`, then just limiting the results... not sure I like that. We should either viewport it properly, or truncate if we exceed a limit at time of `put`.

Reply:
@mofojed  My thought process of this approach was that we could potentially make the limit configurable in the future without losing any underlying history if a user wanted to increase the limit at the expense of slower performance.

Regarding your suggestions:
1. Setting a limit on how much we store would probably be the simplest solution, but someone would need to make a call on what that limit needs to be
2. For "viewport it properly", I think this would mean we need to define a large window of data to minimize the number of queries required during scrolling. e.g. we could request a few hundred extra rows on top and bottom but then only re-fetch after user has scrolled within a threshold close to the edge of the buffer. Unfortunately I think this will mean that scrolling for filtered data is going to suffer since it seems that any query on a large dataset is going to be slow.

Reply:
Having the limit configurable should really be our last option - we don't want to present unnecessary options to the user. I think I'll just pick an arbitrary number of 2500 commands and that should be plenty. 
We want persistence to be server side at some point anyway, so if we do `allDocs()` with a low threshold of 2000 and upper threshold of 2500 (e.g. once you write the 2500th record, it deletes the top 500 bringing it back down to 2000), then we should be good. Although, even looking into the PouchDB documentation, it doesn't actually delete records, it just adds a tombstone: https://pouchdb.com/api.html#delete_document

So I wonder if deleting even helps with performance, or if you need to compact afterwards... How are you currently testing this/with how many records? Still need to include the testing steps in the PR description.

Side note, read this on IndexedDB performance (from someone who used to work on PouchDB): https://nolanlawson.com/2021/08/22/speeding-up-indexeddb-reads-and-writes/

Reply:
I've tested with 20,000 records so far. I plan to add some more formal details to the top of the PR. Was waiting until I finalize the code